README

the aEGiS PHP nano web server
=============================

Copyright (C) 2002-2005 Vincent Negrier aka. sIX <six@aegis-corp.org>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


What ??! 

Nanoweb is an HTTP server written in PHP, designed to be small, secure, 
and extensible.

The main features are :

- HTTP/1.1 compliance
- Powerful and easy configuration
- Modular architecture
- FastCGI, CGI and Server side includes support
- Name and port based virtual hosts
- Access control lists
- htpasswd, MySQL, PostgreSQL and LDAP authentication support
- Themes for server generated content
- Apache compatible log format, MySQL logging
- Directory browsing
- inetd support and SSL via external helpers
- Denial of Service protection
- Proxy Server extension
- Filters and gzip support
- RBL support (mail-abuse.org)
- Extension Protocols (request methods) support
- ... and a lot more


Why ? 

Why did I do it ? I found the idea of using a language mostly aimed at web 
scripting to write a web server was quite funny. That's all, really ...

Why should you use it ?

It's light, fast and robust.
Also being written in PHP makes it very easily extensible with custom modules,
and secure because - running inside the Zend scripting engine - it's immune
to a number of common security threats (format strings, buffer overflows, ...)


How ? 

Read INSTALL, and follow the simple installation steps.

Interesting readings after install is done :

- User Manual (http://localhost/manual or http://nanoweb.si.kz/manual)
- The 'docs' directory contains some brief infos about individual parts of the
  server, modules, and some howtos.
- The 'ChangeLog' file contains the list of all added features and this could
  be useful.
- Default configuration files are well commented and reading them should help
  you getting started too.

For any other concerns about nanoweb, please consult the general or support
forums at <http://forums.t0x.net/?c=2> or mail the author.



ChangeLog
2.2.9 - 2008/06/10

- Fixed socket_select() call semantics for PHP >= 5.2.6 compatibility
- Fixed a small bug in mod_nospam which could lead to undetected spam
- Fixed parameters passed to tail in install-sh


2.2.8 - 2006/08/10

- Added options in nanoctl for managing the internal address block list
- Added helper module mod_blockadm for the new nanoctl options (block/unblock)
- Added mod_nospam, a generic GET and POST content scanner based on regex
- Fixed a very annoying bug with mod_cgi and env variables on win32
- Fixed response for requests having an "Expect: 100-continue" header
- Fixed detection of duplicate request headers and appropriate error message
- Fixed a small bug with duplicate error pages in single process mode 


2.2.7 - 2005/05/10

- Added restarting of logger processes when receiving SIGHUP (nanoctl reload)
- Fixed automatic directory redirection to handle query strings correctly
- Fixed "FATAL: socket bind failed" when restarting with open connections
- Fixed FastCGI packet encoding bug in mod_fcgi (Martin Schmid)
- Fixed erroneous debug logging when using the AuthLocation directive
- Fixed reloading not re-resolving symbolic links in DocumentRoot directives
- Fixed default windows config files to set "KeepAlive = 0"
- Optimization of FastCGI packet encoding and decoding in mod_fcgi


2.2.6 - 2005/01/09

- Added detection of phpbb exploits and other HTTP pest by mod_worms
- Fixed OS detection so that Cygwin is no longer treated as unix
- Fixed access files not being read in allowed subdirectories for symlinks
- Fixed bug when using multiple LoadTheme and/or ServerTheme in the same scope


2.2.5 - 2004/10/16

- Added forward dns lookup on resolved hostnames to validate the responses
- Added mod_access_referer (see RefererCheck and RefererAllow directives)
- Added display of server processes uptimes in "nanoctl status who"
- Added warning message when not using "KeepAlive=0" in single process mode
- Fixed a small bug in the handling of null parser object responses
- Fixed the "Max execution time exceeded" problem in single process mode


2.2.4 - 2004/08/15

- Added sorting of directories in mod_fb using the chosen file order
- Fixed the access_query() function on merged properties (PHP5 compat)
- Fixed reopen of standard input in daemon initialization code


2.2.3 - 2004/05/14

- Added "AuthLocation" directive to restrict authentication to given path(s)
- Added mod_auth_subversion for authentication against svn repositories (Jimbo)
- Fixed REMOTE_USER server var that was not set on authenticated requests
- Fixed core parser reading loop (may have caused truncation with mod_fcgi)
- Fixed a variable access bug in modules_init()
- Fixed OS detection so that "Darwin" is no longer treated as Win*


2.2.2 - 2003/11/10

- Added checking of requested URL against known device names on win32
- Added checks and adequate responses for bad request protocols and headers
- Added reporting of process memory usage (when available) to mod_status
- Changed all "<?" occurences in nanoweb core and modules to "<?php" for
  compatibility with "short_open_tag = Off" in php config


2.2.1 - 2003/07/25

- Added spamtrap-friendly addresses generation in mod_emailprotect
- Fixed two small bugs in the xml output of mod_status (nanoctl status *-xml)
- Minor optimizations of the initialization code


2.2.0 - 2003/06/09

- Added e-mail address protection filter module mod_emailprotect (Mario)
- Added authentication module for PEAR::DB, ADOdb and dbx_() supported RDBMS
- Added support for NCSA Common Log Format and CLF-with-vhost in mod_stdlog
- Added ftp-like anonymous authentication module mod_auth_anonymous (Mario)
- Added support for HTTP "Expect: 100-continue" request header
- Added output of an "Allow" response header for 405 errors (forbidden method)
- Added blocking of incoming connections by IP addresses (see "BlockIPAddr")
- Added "StaticDisablePartial" directive to disallow serving partial content
- Added support for dynamic IP blocking to mod_worms, mod_dosevasive, mod_ac
- Added download bandwidth limiting module mod_throttle (Mario)
- Added support for the "LoadTheme" directive in access files
- Added child process spawn timestamp in master scoreboard
- Added setting of socket send and receive buffers according to SCK_*_SIZE
- Added a "blocking" server log level for IP addresses blocking services
- Added an optional shutdown() callback method for modules
- Fixed loggers watcher, any logger dying unexpectedly is now auto-respawned
- Cleaning and optimization of the IPC functions (now use SOCK_DGRAM pairs)
- Changed "LoadTheme" so that it also sets the current theme to the loaded one


2.1.3 - 2003/04/28

- Fixed server vars cleaning in mod_cgi when running in single process mode
- Fixed a small addslashes() related bug in contrib/mkhugenanoweb.php
- Fixed the "Content-Length" header for HEAD requests on filtered resources
- Fixed handling of connections dropped before initial socket setup
- Fixed the multiline output of server log events with different levels
- Fixed logging of request with double quotes in uri, user agent, or referer
- Fixed propagation of configuration to virtual hosts sections for directives
  located after the vhost declaration


2.1.2 - 2003/04/02

- Added to nanoctl a "configtest" option to check for configuration errors
- Added "FBDateFormat" conf/access directive to mod_fb
- Fixed "405 Method Not Allowed" when POSTing to a script in a CGIScriptDir
- Fixed a small bug with the access files loader when running on Win32
- Fixed and cleaned some ugly code in the user directories handler
- Fixed advise/redirect URL in mod_mispell if inside an aliased or user dir
- Fixed keep alive responses when running in single process mode
- Fixed handling of reqs for different vhosts in the same keep alive session
- Fixed the internal stats counter numbers in single process mode
- Fixed a small display bug in nanoctl when nanoweb starts with warnings

modules/mod_fcgi.php
<?php

/*

Nanoweb FastCGI support module
==============================

Copyright (C) 2002-2003 Vincent Negrier aka. sIX <six@aegis-corp.org>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

*/

define(FCGI_VERSION_1, 1);

define(FCGI_BEGIN_REQUEST, 1);
define(FCGI_ABORT_REQUEST, 2);
define(FCGI_END_REQUEST, 3);
define(FCGI_PARAMS, 4);
define(FCGI_STDIN, 5);
define(FCGI_STDOUT, 6);
define(FCGI_STDERR, 7);
define(FCGI_DATA, 8);
define(FCGI_GET_VALUES, 9);
define(FCGI_GET_VALUES_RESULT, 10);

class mod_fcgi {

	function mod_fcgi() {

		$this->modtype="parser_FCGI";
		$this->modname="FastCGI support";
	
	}

	function build_fcgi_packet($type, $content) {

		$clen=strlen($content);
				
		$packet=chr(FCGI_VERSION_1);
		$packet.=chr($type);
		$packet.=chr(0).chr(1); // Request id = 1
		$packet.=chr((int)($clen/256)).chr($clen%256); // Content length
		$packet.=chr(0).chr(0); // No padding and reserved
		$packet.=$content;

		return($packet);
	
	}

   function build_fcgi_nvpair($name, $value) {

		$nlen = strlen($name);
		$vlen = strlen($value);
			 
		if ($nlen < 128) {
		   
			$nvpair = chr($nlen);

		} else {
		   
			$nvpair = chr(($nlen >> 24) | 0x80) . chr(($nlen >> 16) & 0xFF) . chr(($nlen >> 8) & 0xFF) . chr($nlen & 0xFF);

		}

		if ($vlen < 128) {
		   
			$nvpair .= chr($vlen);

		} else {
		   
			$nvpair .= chr(($vlen >> 24) | 0x80) . chr(($vlen >> 16) & 0xFF) . chr(($vlen >> 8) & 0xFF) . chr($vlen & 0xFF);

		}

		return $nvpair . $name . $value;
     
	} 
	
	function decode_fcgi_packet($data) {

		$ret["version"]=ord($data{0});
		$ret["type"]=ord($data{1});
		$ret["length"]=(ord($data{4}) << 8)+ord($data{5});
		$ret["content"]=substr($data, 8, $ret["length"]);

		return($ret);
	
	}

	function parser_open($args, $filename, &$rq_err, &$cgi_headers) {

		global $conf, $add_errmsg;
		
		// Connect to FastCGI server

		$fcgi_server=explode(":", $args);

		if (!$this->sck=fsockopen($fcgi_server[0], $fcgi_server[1], $errno, $errstr, 5)) {

			$rq_err=500;
			$tmperr="mod_fcgi: unable to contact application server ($errno : $errstr).";
			$add_errmsg.=($tmperr."<br><br>");
			techo("WARN: ".$tmperr, NW_EL_WARNING);
			return (false);

		}

		// Begin session
		
		$begin_rq_packet=chr(0).chr(1).chr(0).chr(0).chr(0).chr(0).chr(0).chr(0);
		fwrite($this->sck, $this->build_fcgi_packet(FCGI_BEGIN_REQUEST, $begin_rq_packet));

		// Build params
		
		$fcgi_params_packet.=$this->build_fcgi_nvpair("GATEWAY_INTERFACE", "FastCGI/1.0");
		$nsv=nw_server_vars();
		if ($conf["global"]["fcgifilterpathinfo"][0]) unset($nsv["PATH_INFO"]);
		foreach($nsv as $key=>$var) $fcgi_params_packet.=$this->build_fcgi_nvpair($key, $var);

		if ($rq_hdrs=$GLOBALS["htreq_headers"]) foreach ($rq_hdrs as $key=>$val) $fcgi_params_packet.=$this->build_fcgi_nvpair("HTTP_".str_replace("-", "_", $key),$val);
		
		if ($GLOBALS["http_action"]=="POST" && $GLOBALS["htreq_content"]) {
		
			$fcgi_params_packet.=$this->build_fcgi_nvpair("CONTENT_TYPE", $rq_hdrs["CONTENT-TYPE"]);
			$fcgi_params_packet.=$this->build_fcgi_nvpair("CONTENT_LENGTH", $rq_hdrs["CONTENT-LENGTH"]);

			$stdin_content=$GLOBALS["htreq_content"];
			
		} else $stdin_content="";

		// Send params
		
		fwrite($this->sck, $this->build_fcgi_packet(FCGI_PARAMS, $fcgi_params_packet));
		fwrite($this->sck, $this->build_fcgi_packet(FCGI_PARAMS, ""));
		
		// Build and send stdin flow

		if ($stdin_content) fwrite($this->sck, $this->build_fcgi_packet(FCGI_STDIN, $stdin_content));
		fwrite($this->sck, $this->build_fcgi_packet(FCGI_STDIN, ""));

		// Read answers from fastcgi server

		$content="";

		while (($p1=strpos($content, "\r\n\r\n"))===false) {

			$tmpp=$this->decode_fcgi_packet($packet=fread($this->sck, 8));
			$tl=$tmpp["length"]%8;
			$tadd=($tl?(8-$tl):0);
			$resp=$this->decode_fcgi_packet($packet.fread($this->sck, $tmpp["length"]+$tadd));

			if ($valid_pck=($resp["type"]==FCGI_STDOUT || $resp["type"]==FCGI_STDERR)) $content.=$resp["content"];

			if ($resp["type"]==FCGI_STDERR) techo("WARN: mod_fcgi: app server returned error : '".$resp["content"]."'", NW_EL_WARNING);

		}

		if (feof($this->sck)) $this->peof=true;
		
		if ($p1) {
		
			$headers=explode("\n", trim(substr($content, 0, $p1)));
			$content=substr($content, $p1+4);

		}

		$GLOBALS["http_resp"]="";
		
		$cnh=access_query("fcginoheader");
		
		foreach ($headers as $s) if ($s=trim($s)) {

			if (substr($s, 0, 5)=="HTTP/") {

				$hd_key="STATUS";
				strtok($s, " ");
			
			} else {
			
				$hd_key=strtok($s, ":");

			}

			$hd_val=trim(strtok(""));
			$hku=strtoupper($hd_key);
			
			if ($cnh) foreach ($cnh as $nohdr) if ($hku==strtoupper($nohdr)) $hd_key="";
			
			if ($hd_key) {
			
				if ($hku=="SET-COOKIE") {
					
					$cgi_headers["cookies"][]=$hd_val;

				} else {

					$cgi_headers[$hd_key]=$hd_val;

				}

			}
		
		}

		$this->parsed_output=$content;

	}

	function parser_get_output() {

		if (!$this->peof && !$this->parsed_output) {
		
			$tmpp=$this->decode_fcgi_packet($packet=fread($this->sck, 8));
			$tl=$tmpp["length"]%8;
			$tadd=($tl?(8-$tl):0);
			$resp=$this->decode_fcgi_packet($packet.fread($this->sck, $tmpp["length"]+$tadd));

			if ($valid_pck=($resp["type"]==FCGI_STDOUT || $resp["type"]==FCGI_STDERR)) {
				
				$content.=$resp["content"];

			} else {

				$this->peof=true;				
			
			}

			if ($resp["type"]==FCGI_STDERR) techo("WARN: mod_fcgi: app server returned error : '".$resp["content"]."'", NW_EL_WARNING);

		}

		if ($this->parsed_output) {

			$content=$this->parsed_output;
			$this->parsed_output="";
		
		}

		return($content);
	
	}
	
	function parser_eof() {

		return($this->peof);
	
	}
	
	function parser_close() {

		$this->peof=false;
		fclose($this->sck);

	}

}

?>
